cmds/core/ping/ping.go

```diff
@@ -44,145 +46,83 @@ var (
+	wb, err := wm.Marshal(nil)
+	if err != nil {
+		return "", fmt.Errorf("icmp.Message.Marshal failed: %v", err)
 	}
 
-	// Get ICMP Echo Reply
-	c.SetDeadline(time.Now().Add(waitFor))
-	rmsg := make([]byte, *packetSize+256)
-	before := time.Now()
-	amt, err := c.Read(rmsg[:])
+	startTime := time.Now()
+	_, err = conn.WriteTo(wb, &net.IPAddr{IP: net.ParseIP(host)})
 	if err != nil {
-		return "", fmt.Errorf("read failed: %v", err)
-	}
-	latency := time.Since(before)
-	if !net6 {
-		rmsg = rmsg[ICMP_ECHO_REPLY_HEADER_IPV4_OFFSET:]
-	}
-	if net6 {
-		if rmsg[0] != ICMP6_TYPE_ECHO_REPLY {
-			return "", fmt.Errorf("bad ICMPv6 echo reply type, got %d, want %d", rmsg[0], ICMP6_TYPE_ECHO_REPLY)
-		}
-	} else {
-		if rmsg[0] != ICMP_TYPE_ECHO_REPLY {
-			return "", fmt.Errorf("bad ICMP echo reply type, got %d, want %d", rmsg[0], ICMP_TYPE_ECHO_REPLY)
-		}
+		return "", fmt.Errorf("conn.Write failed: %v", err)
 	}
-	cks := binary.BigEndian.Uint16(rmsg[2:])
-	binary.BigEndian.PutUint16(rmsg[2:], 0)
-	// only validate the checksum for IPv4. For IPv6 this *should* be done by the
-	// TCP stack (and do we need to validate the checksum anyway?)
-	if !net6 && cks != cksum(rmsg) {
-		return "", fmt.Errorf("bad ICMP checksum: %v (expected %v)", cks, cksum(rmsg))
-	}
-	id := binary.BigEndian.Uint16(rmsg[4:])
-	seq := binary.BigEndian.Uint16(rmsg[6:])
-	rseq := uint64(id)<<16 + uint64(seq)
-	if rseq != i {
-		return "", fmt.Errorf("wrong sequence number %v (expected %v)", rseq, i)
+
+	rb := make([]byte, 1500)
+	n, _, err := conn.ReadFrom(rb)
+	if err != nil {
+		return "", fmt.Errorf("conn.Read failed: %v", err)
 	}
 
-	return fmt.Sprintf("%d bytes from %v: icmp_seq=%v, time=%v", amt, host, i, latency), nil
-}
+	latency := time.Since(startTime)
 
-func ping(host string) error {
-	if *packetSize < 8 {
-		return fmt.Errorf("packet size too small (must be >= 8): %v", *packetSize)
+	msg, err := icmp.ParseMessage(ipv4.ICMPTypeEchoReply.Protocol(), rb[:n])
+	if err != nil {
+		log.Fatal(err)
 	}
 
-	interval := time.Duration(*intv)
-	p := New()
-	// ping needs to run forever if count is not specified, so default value is MaxUint64
-	waitFor := time.Duration(*wtf) * time.Millisecond
-	for i := uint64(0); i < *iter; i++ {
-		msg, err := p.ping1(*net6, host, i+1, waitFor)
-		if err != nil {
-			return fmt.Errorf("ping failed: %v", err)
-		}
-		if *audible {
-			msg = "\a" + msg
-		}
-		log.Print(msg)
-		time.Sleep(time.Millisecond * interval)
+	echoReply, ok := msg.Body.(*icmp.Echo)
+	if !ok {
+		return "", fmt.Errorf("got %+v; want echo reply", msg)
 	}
 
-	return nil
+	if echoReply.ID != os.Getpid()&0xffff {
+		return "", fmt.Errorf("got id %v; want %v", echoReply.ID, os.Getpid()&0xffff)
+	}
+	if echoReply.Seq != i {
+		return "", fmt.Errorf("got seq %v; want %v", echoReply.Seq, i)
+	}
+
+	return fmt.Sprintf("%d bytes from %v: icmp_seq=%v, time=%v", n, host, i, latency), nil
 }
 
 func main() {
+	_ = icmp.Echo{}
```

Comment:
This confused me, why was it needed? Is it to force inclusion of icmp? You can do that on the import line if so:
import _ "foo.bar"

Reply:
this is junk, I just want to see if I have free ride with testing it with raw socket under our CI, but I will probably need to provide testing interface :)

Reply:
A vmtest would be appropriate here. raw socket support will exist there